pkg/hs/urbit-king/lib/Urbit/King/App.hs

```diff
@@ -73,36 +76,50 @@ instance HasKingId KingEnv where
 
 -- Running KingEnvs ------------------------------------------------------------
 
-runKingEnvStderr :: Bool -> RIO KingEnv a -> IO a
-runKingEnvStderr verb inner = do
+runKingEnvStderr :: Bool -> LogLevel -> RIO KingEnv a -> IO a
+runKingEnvStderr verb lvl inner = do
   logOptions <-
-    logOptionsHandle stderr verb <&> setLogUseTime True <&> setLogUseLoc False
+    logOptionsHandle stderr verb
+      <&> setLogUseTime True
+      <&> setLogUseLoc False
+      <&> setLogMinLevel lvl
 
   withLogFunc logOptions $ \logFunc -> runKingEnv logFunc logFunc inner
 
-runKingEnvLogFile :: Bool -> RIO KingEnv a -> IO a
-runKingEnvLogFile verb inner = withLogFileHandle $ \h -> do
-  logOptions <-
-    logOptionsHandle h verb <&> setLogUseTime True <&> setLogUseLoc False
-  stderrLogOptions <-
-    logOptionsHandle stderr verb <&> setLogUseTime False <&> setLogUseLoc False
-
-  withLogFunc stderrLogOptions $ \stderrLogFunc -> withLogFunc logOptions
-    $ \logFunc -> runKingEnv logFunc stderrLogFunc inner
-
-withLogFileHandle :: (Handle -> IO a) -> IO a
-withLogFileHandle act = do
-  home <- getHomeDirectory
-  let logDir = home </> ".urbit"
-  createDirectoryIfMissing True logDir
-  withFile (logDir </> "king.log") AppendMode $ \handle -> do
+runKingEnvLogFile :: Bool -> LogLevel -> Maybe FilePath -> RIO KingEnv a -> IO a
+runKingEnvLogFile verb lvl fileM inner = do
+  logFile <- case fileM of
+    Just f  -> pure f
+    Nothing -> defaultLogFile
+  withLogFileHandle logFile $ \h -> do
+    logOptions <-
+      logOptionsHandle h verb
+        <&> setLogUseTime True
+        <&> setLogUseLoc False
+        <&> setLogMinLevel lvl
+    stderrLogOptions <-
+      logOptionsHandle stderr verb
+        <&> setLogUseTime False
+        <&> setLogUseLoc False
+        <&> setLogMinLevel lvl
+
+    withLogFunc stderrLogOptions $ \stderrLogFunc -> withLogFunc logOptions
+      $ \logFunc -> runKingEnv logFunc stderrLogFunc inner
+
+withLogFileHandle :: FilePath -> (Handle -> IO a) -> IO a
+withLogFileHandle f act =
+  withFile f AppendMode $ \handle -> do
     hSetBuffering handle LineBuffering
     act handle
 
+defaultLogFile :: IO FilePath
+defaultLogFile = do
+  logDir <- getAppUserDataDirectory "urbit"
```

Comment:
Consider `getXdgDirectorySource XdgCache "urbit"`. I was really surprised when it dumped log files in a new `~/.urbit/` originally.

Reply:
I saw this Xdg business in the docs, but I didn't really understand it, so I passed over it. I'm happy to follow your suggestion though.

Reply:
Okay I understand it now.